File: backend/app/routers/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models import Project
from app.schemas.projects.project import ProjectCreate, ProjectUpdate, ProjectOut

router = APIRouter(prefix="/projects", tags=["Projects"])


@router.get("/", response_model=list[ProjectOut])
def get_all_projects(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return db.query(Project).offset(skip).limit(limit).all()


@router.get("/{project_id}", response_model=ProjectOut)
def get_project(project_id: int, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


@router.get("/company/{company_id}", response_model=list[ProjectOut])
def get_by_company(company_id: int, db: Session = Depends(get_db)):
    return db.query(Project).filter(Project.company_id == company_id).all()


@router.post("/", response_model=ProjectOut, status_code=201)
def create_project(data: ProjectCreate, db: Session = Depends(get_db)):
    project = Project(**data.model_dump())
    db.add(project)
    db.commit()
    db.refresh(project)
    return project


@router.patch("/{project_id}", response_model=ProjectOut)
def update_project(project_id: int, data: ProjectUpdate, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(project, field, value)
    db.commit()
    db.refresh(project)
    return project


@router.delete("/{project_id}", status_code=204)
def delete_project(project_id: int, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    db.delete(project)
    db.commit()

File: frontend/lib/api/projects.ts
/**
 * lib/api/projects.ts
 */

import { apiRequest } from "./client";

export interface ProjectOut {
  id: number;
  name?: string;
  description?: string;
  company_id?: number;
  created_by?: number;
  created_at?: string;
}

export interface ProjectCreate {
  name: string;
  company_id: number;
  created_by: number;
  description?: string;
}

export interface ProjectUpdate {
  name?: string;
  description?: string;
  company_id?: number;
}

export async function getProjects(skip = 0, limit = 100): Promise<ProjectOut[]> {
  return apiRequest<ProjectOut[]>("/projects", { params: { skip, limit } });
}

export async function getProject(projectId: number): Promise<ProjectOut> {
  return apiRequest<ProjectOut>(`/projects/${projectId}`);
}

export async function getProjectsByCompany(companyId: number): Promise<ProjectOut[]> {
  return apiRequest<ProjectOut[]>(`/projects/company/${companyId}`);
}

export async function createProject(data: ProjectCreate): Promise<ProjectOut> {
  return apiRequest<ProjectOut>("/projects", { method: "POST", body: data });
}

export async function updateProject(projectId: number, data: ProjectUpdate): Promise<ProjectOut> {
  return apiRequest<ProjectOut>(`/projects/${projectId}`, { method: "PATCH", body: data });
}

export async function deleteProject(projectId: number): Promise<void> {
  return apiRequest<void>(`/projects/${projectId}`, { method: "DELETE" });
}
